bftengine/src/bftengine/ClientsManager.cpp

```diff
@@ -95,26 +97,33 @@ void ClientsManager::loadInfoFromReservedPages() {
   for (auto const& [clientId, clientIdx] : clientIdToIndex_) {
     const uint32_t firstPageId = clientIdx * reservedPagesPerClient_;
 
-    if (!stateTransfer_->loadReservedPage(resPageOffset() + firstPageId, sizeOfReservedPage_, scratchPage_)) continue;
-
-    ClientReplyMsgHeader* replyHeader = (ClientReplyMsgHeader*)scratchPage_;
-    ConcordAssert(replyHeader->msgType == 0 || replyHeader->msgType == MsgCode::ClientReply);
-    ConcordAssert(replyHeader->currentPrimaryId == 0);
-    ConcordAssert(replyHeader->replyLength >= 0);
-    ConcordAssert(replyHeader->replyLength + sizeof(ClientReplyMsgHeader) <= maxReplySize_);
-
-    // YS TBD: Multiple replies for client batching should be sorted by incoming time
-    auto& repliesInfo = indexToClientInfo_.at(clientIdx).repliesInfo;
-    if (repliesInfo.size() >= maxNumOfReqsPerClient_) deleteOldestReply(clientId);
-    const auto& res = repliesInfo.insert_or_assign(replyHeader->reqSeqNum, MinTime);
-    const bool added = res.second;
-    LOG_INFO(CL_MNGR, "Added/updated reply message" << KVLOG(clientId, replyHeader->reqSeqNum, added));
-
+    // load all replies into the map of replies one by one
+    for (int i = 0; i < maxNumOfReqsPerClient_; i++) {
+      if (!stateTransfer_->loadReservedPage(
+              resPageOffset() + firstPageId + i, sizeOfReservedPage_, scratchPage_.data() + i * sizeOfReservedPage_))
+        continue;
+
+      auto replyHeader = (ClientReplyMsgHeader*)(scratchPage_.data() + i * sizeOfReservedPage_);
+      auto replyPtr = std::make_shared<ClientReplyMsg>(myId_, replyHeader->replyLength);
+      ConcordAssert(replyHeader->msgType == 0 || replyHeader->msgType == MsgCode::ClientReply);
+      ConcordAssert(replyHeader->currentPrimaryId == 0);
+      ConcordAssert(replyHeader->replyLength >= 0);
+      ConcordAssert(replyHeader->replyLength + sizeof(ClientReplyMsgHeader) <= maxReplySize_);
+
+      auto& repliesInfo = indexToClientInfo_.at(clientIdx).repliesInfo;
+      if (repliesInfo.size() >= maxNumOfReqsPerClient_) deleteOldestReply(clientId);
+      const auto& res = repliesInfo.insert({replyHeader->reqSeqNum, replyPtr});
+      const bool added = res.second;
+      LOG_INFO(CL_MNGR, "Added/updated reply message" << KVLOG(clientId, replyHeader->reqSeqNum, added));
+    }
+    // get the newest reply from the map of replies.
+    auto lastReply = indexToClientInfo_.at(clientIdx).repliesInfo.cend();
```

Comment:
it seems to be a problem:`cend` returns an element after the last, i.e. it is not the actual last element.
We can either change the sorting order to descending (@yuliasherman can probably advise here) or get the last item with something like this:
```
std::prev(indexToClientInfo_.at(clientIdx).repliesInfo.end());
```

Reply:
You are right thanks, I will go with rbegin() - the reverse iterator.

Reply:
@Liorgal28, perfecto!
Thank you